cloudinit/config/cc_kernel_modules.py

```diff
@@ -0,0 +1,373 @@
+                            )
+    return errors
+
+
+def supplemental_schema_validation(kernel_module: dict):
+    """Validate user-provided kernel_modules option values.
+
+    This function supplements flexible jsonschema validation with specific
+    value checks to aid in triage of invalid user-provided configuration.
+
+    @param kernel_module: Dictionary.
+
+    @raises: ValueError describing invalid values provided.
+    """
+    errors = []
+    missing = REQUIRED_KERNEL_MODULES_KEYS.difference(
+        set(kernel_module.keys())
+    )
+    if missing:
+        keys = ", ".join(sorted(missing))
+        errors.append(f"Missing required kernel_modules keys: {keys}")
+
+    for key, value in sorted(kernel_module.items()):
+        if key == "name":
+            if not isinstance(value, str):
+                errors.append(
+                    "Expected a string for kernel_modules:"
+                    f"{key}. Found {value}."
+                )
+        elif key == "load":
+            if not isinstance(value, bool):
+                errors.append(
+                    "Expected a boolean for kernel_modules:"
+                    f"{key}. Found {value}."
+                )
+        elif key == "persist":
+            errors += persist_schema_validation(kernel_module[key])
+
+    if errors:
+        raise ValueError(
+            f"Invalid kernel_modules configuration:{NL}{NL.join(errors)}"
+        )
+
+
+def prepare_module(module_name: str):
+    """Prepare kernel module to be loaded on boot.
+
+    This function appends a kernel module to a file
+    (depends on operating system) for loading a module on boot.
+
+    @param module_name: string.
+
+    @raises: RuntimeError when write operation fails.
+    """
+
+    try:
+        LOG.debug("Appending kernel module %s", module_name)
+        util.write_file(
+            DEFAULT_CONFIG["load"]["path"],
+            module_name + NL,
+            DEFAULT_CONFIG["load"]["permissions"],
+            omode="a",
+        )
+    except Exception as e:
+        raise RuntimeError(
+            f"Failure appending kernel module '{module_name}' to file "
+            f'{DEFAULT_CONFIG["load"]["path"]}:{NL}{str(e)}'
+        ) from e
+
+
+def enhance_module(module_name: str, persist: dict, unload_modules: list):
+    """Enhances a kernel module's behaviour
+
+    This function appends specific settings and options
+    for a kernel module, which will be applied when the kernel
+    module gets loaded.
+
+    @param module_name: string.
+    @param persist: Dictionary
+    qparam unload_modules: list
+
+    @raises RuntimeError when write operation fails.
+    """
+
+    for (key, value) in persist.items():
+        LOG.debug(
+            "Enhancing kernel module %s with %s:%s", module_name, key, value
+        )
+        entry = f"{key} {module_name} {value}"
+        # softdep special case
+        if key == "softdep":
+            entry += " ".join({key["pre"]}) + " ".join({key["post"]})
```

Comment:
This entry doesn't appear to be constructed in the format expected by modprobe.d.
Shouldn't it be of the format: `softdep wireguard pre: <all_pre_module_name> post: <all_post_module_names>` ?

The way this entry is appended , I don't see either `pre:` or `post:` definitions in the output.



you have done a fair job establishing failure conditions and error handling in the tests. But, I think we need  to see a happy path test exercising each function and asserting the correct output is being written to the right DEFAULT_CONFIG...<path>.